File: src/py_pkg/py_pkg/sum_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial

"""
ros2  service call  /sum_service  example_interfaces/srv/AddTwoInts  "{'a': 4, 'b': 7}"

ros2 run <pkg_name> <node_name> –ros-args -p <param1_name>:=<value> -p <param2_name>:=<value>

"""

class SumClientNode(Node):
    def __init__(self):
        super().__init__('sum_client_node')
        self.get_logger().info('Sum Client Python node has been created')

        #ROS2 Part 6 -----------------------------------
        #a_ = 6
        #b_ = 7

        # declare parameters
        self.declare_parameter('a', 0) #default
        self.declare_parameter('b', 0) #default
        
        #get parameters
        a_ = self.get_parameter('a').get_parameter_value().integer_value
        b_ = self.get_parameter('b').get_parameter_value().integer_value
        
        self.get_logger().info(f'El valor de a es: {a_}')
        self.get_logger().info(f'El valor de b es: {b_}')
        #-----------------------------------
        
        self.call_sum_server(a_, b_)
    
    def call_sum_server(self, a, b):
        #client = self.create_client(<ServiceType>, <ServiceName>)
        client = self.create_client(AddTwoInts, 'sum_service')
        #Luego de crear el cliente, debemos esperar hasta que el Servicio esté activo, es decir, el nodo Servidor esté ejecutándose.
        while not client.wait_for_service(1.0):
            self.get_logger().info('Waiting for the Server...')

        # create request
        request = AddTwoInts.Request()
        request.a = a
        request.b = b


        #send request asynchronously
        future = client.call_async(request)

        # La llamada asincrónica devolverá un objeto de la clase rclpy.task.Future, que indicará si la llamada 
        # y la respuesta han finalizado o no, a través de la función add_done_callback en la siguiente línea.
        future.add_done_callback(partial(self.sum_service_callback, a=a, b=b))

    def sum_service_callback(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(f'{a} + {b} = {response.sum}')
        except Exception as e:
            self.get_logger().info(f'Service call failed {e}')

def main(args=None):
    rclpy.init(args=args)
    node = SumClientNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__=='__main__':
    main()

File: src/cpp_pkg/src/sum_client.cpp
#include "example_interfaces/srv/add_two_ints.hpp"
#include "rclcpp/rclcpp.hpp"

class SumClientNode : public rclcpp::Node {
public:
  SumClientNode() : Node("sum_client_node") {
    RCLCPP_INFO(this->get_logger(), "Service client C++ node has been created");

    // Parameters for AddTwoInts
    a_ = 4;
    b_ = 7;
    threads_.push_back(
        std::thread(std::bind(&SumClientNode::CallSumServer, this, a_, b_)));
  }
  void CallSumServer(int a, int b) {
    // create client
    client_ =
        this->create_client<example_interfaces::srv::AddTwoInts>("sum_service");
    while (!client_->wait_for_service(std::chrono::seconds(1))) {
      RCLCPP_WARN(this->get_logger(), "Waiting for the Server...");
    }

    // create request message
    auto request =
        std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
    request->a = a;
    request->b = b;

    // future = client.call_async
    auto future = client_->async_send_request(request);
    try {
      auto response = future.get();
      RCLCPP_INFO(this->get_logger(), "%d + %d = %d", a, b, (int)response->sum);
    } catch (const std::exception &e) {
      RCLCPP_ERROR(this->get_logger(), "Service call failed");
    }
  }

private:
  rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client_;
  // La principal diferencia con la implementación de Python se encuentra en
  // Line14 y Line41, donde declaramos  std::vector threads_, un vector de hilos
  // al que agregamos todas las llamadas a funciones de CallSumServer, que se
  // procesarán de forma asincrónica en hilos separados.
  std::vector<std::thread> threads_;
  int a_, b_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SumClientNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
